File: inspirehep/modules/deposit/static/js/deposit/data_sources/doi.js
/*
 * This file is part of INSPIRE.
 * Copyright (C) 2014 CERN.
 *
 * INSPIRE is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * INSPIRE is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
 *
 * In applying this licence, CERN does not waive the privileges and immunities
 * granted to it by virtue of its status as an Intergovernmental Organization
 * or submit itself to any jurisdiction.
 */

define(function(require, exports, module) {
  var DataSource = require("js/deposit/data_sources/data_source");
  var DataMapper = require("js/deposit/mapper");
  var doiSource = new DataSource({

    id: 'doi',
    name: 'DOI',
    service: 'the DOI',
    url: '/doi/search?doi=',

    mapper: new DataMapper({

      common_mapping: function(data) {

        var journal;

        if ("container-title" in data) {
          for (var i = 0; i < data['container-title'].length; i++) {
            if (i === data['container-title'].length - 1) {
              journal = data['container-title'][i];
            } else {
              journal = data['container-title'][i] + ', ';
            }
          }
        }

        var doi_obj = {
          journal_title: journal,
          isbn: data.isbn,
          page_range_article_id: data.page,
          year: data.issued['date-parts'][0][0],
          issue: data.issue,
          authors: data.author,
          volume: data.volume,
          url: data.URL
        };

        // filtering the object out of undefined values
        for (var idx in doi_obj) {
          if (doi_obj[idx] === null || doi_obj[idx] === undefined || doi_obj[idx].length === 0) {
            delete doi_obj[idx];
          }
        }

        return doi_obj;
      },

      special_mapping: {
        thesis: function(data) {
          return {
            title: data.volume_title
          };
        },
        article: function(data) {
          var extra_mapping = {};
          if (data['title'].length > 0) {
            extra_mapping = {
              title: data.title,
              title_crossref: data.title
            }
          }
          return extra_mapping;
        }
      },

      extract_contributor: function(author) {

        return {
          name: author.family + ', ' + author.given,
          affiliation: ''
        };
      }
    })
  });
  module.exports = doiSource;
});

File: inspirehep/modules/records/serializers/schemas/json/literature/common/doi.py
# -*- coding: utf-8 -*-
#
# This file is part of INSPIRE.
# Copyright (C) 2014-2018 CERN.
#
# INSPIRE is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# INSPIRE is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
#
# In applying this license, CERN does not waive the privileges and immunities
# granted to it by virtue of its status as an Intergovernmental Organization
# or submit itself to any jurisdiction.

from __future__ import absolute_import, division, print_function

from marshmallow import Schema, post_dump, fields


class DOISchemaV1(Schema):
    material = fields.Raw()
    value = fields.Raw()

    @post_dump(pass_many=True)
    def filter(self, data, many):
        if many:
            return self.remove_duplicate_doi_values(data)
        return data

    @staticmethod
    def remove_duplicate_doi_values(dois):
        taken_doi_values = set()
        unique_dois = []
        for doi in dois:
            doi_value = doi.get('value')
            if doi_value not in taken_doi_values:
                taken_doi_values.add(doi_value)
                unique_dois.append(doi)
        return unique_dois
